Allow setting file timestamps from plugins (#1608)

warrkan opened the pull request:
Addressed ticket #1608. Implemented possibility to set modification times for container files and extra dir files.
The modification time value can be:

- epoch_plus_second - current behavior, modification time will be trimmed to epoch + 1 second
- keep_original - original modification time will be kept
- date in ISO 8601 format, e.g 2019-06-18T12:00:00Z

Unfortunately I was not able to run all integration tests successfully because of verification errors in BuildImageMojoIntegrationTest.

chanseokoh commented on jib-core/src/main/java/com/google/cloud/tools/jib/api/KeepOriginalModificationTimeProvider.java:
```diff
@@ -0,0 +1,26 @@
+import java.time.Instant;
+
+/** Modification time provider which returns original file modification time */
+public class KeepOriginalModificationTimeProvider implements ModificationTimeProvider {
```
I'll need to decide whether we want to have this in jib-core. Maybe in jib-plugins-common. Particularly, the method throwing a runtime exception is not an ideal behavior as a library implementation.

warrkan commented:
Thanks @chanseokoh for quick response and a review itself. Yes, probably having possibility to specify modification times for separate files from extra dirs is over-engineering and should be simplified, there was no actual need in it. Anyway, will be waiting for further comment and updates.

chanseokoh commented:
Hi @warrkan, we had a discussion, and at least as an initial implementation, we think we should have a single config that uniformly applies to all the files including the extra directories. If we find a use case for the need to individually customizing mod times of extra directory files, we can implemented that later. For now, we don't see the usefulness of that, like you said you don't need it. So, let's take out anything related to the extra directory customization. Then you can start addressing my other comments.

warrkan commented:
@chanseokoh I addressed your comment about extra configuration. Travis build failed, but at first glance everything seems ok. Can you please take a look?

chanseokoh commented:
@warrkan thanks a lot!

Travis passed. (Someone restarted the build?) Anyways, I guess you can continue working on other open review comments? If you have any doubt, questions, or comment, feel free to update.

warrkan commented:
@chanseokoh I addressed remaining comments. The only open topic left (as I can see) is to move provider implementations out of jib-core.

chanseokoh commented:
Thanks @warrkan! We really appreciate it. This feature was one of the popular requests for Jib.

I ran full integration test internally and noticed that there are integration test failures, which will be difficult for you to debug (the full integration test makes use of pushing to an actual GCP project). And there are a few other things left to refactor and clean up, so we think it's better to merge this to a separate branch first. Then we can work on fixing the failures and more refactoring.

Merging this to the `i1608-warrkan-file-timestamps` branch now.

warrkan commented:
@chanseokoh great news! thanks for reviews and assisting. Hope to see it in master soon.
